components/plurals/src/lib.rs

```diff
@@ -155,7 +161,7 @@ pub enum PluralCategory {
     ///
     /// - 0 in Arabic (ar), Latvian (lv)
     /// - 10~20, 30, 40, 50, ... in Latvian (lv)
-    Zero,
+    Zero = 0,
```

Comment:
Thought: For dataful enums I wonder if in general we should reserve 0 for the value "null"; or, in the case of the PluralCategory enum, null is "other" so should "other" be 0?

Reply:
Thought about it, but I didn't know if reordering the variants would be a breaking change or not (because of the `PartialOrd` derive), so I opted for following the specified order.

Reply:
Though, we could also implement `PartialEq` by hand, but it'll be really weird to have `Zero < Other` and ` Zero as u8 > Other as u8`

Reply:
OK, I see, this enum has `derive(Ord)` which somewhat limits our options

My fully detailed thoughts: I think what you did is fine, but I don't see the harm in making a new enum whose purpose is serialization. We then leave this enum as the public-facing Rust enum. Another thing to consider is, the keys in the ZeroMap do not need to be validated; if you have an out-of-range plural category, you just won't ever find that entry in the map, so we can save a little bit of deserialization code and speed by using a plain `u8` as the key. However, that key type does not serialize nicely in JSON, so what we've done in other places is make yet another enum which is a newtype wrapper over the primitive but has nice serialization (usually called UnvalidatedX: see UnvalidatedTinyAsciiStr, UnvalidatedStr, etc). However, the nice part about your proposed solution is that it is the least amount of additional boilerplate code. It just might be tying too many functional use cases together.

Reply:
Ok, then it would be:
- A new `RawPluralCategory` enum inside `provider` for serialization and deserialization purposes with `RawPluralCategory::Other = 0` and a conversion from `PluralCategory` to `PluralCategory`. This will be the new value type of the map.
- A new `UnvalidatedPluralCategory(u8)` with custom `Serialize` and `Deserialize` impls that de/serializes and serializes from/to a string if the de/serializer is human readable, which will be the key type of the map. Also a conversion from `RawPluralCategory` to `UnvalidatedPluralCategory`.

Is that correct?

Also, should both types be `#[doc(hidden]`? Or should I just mark them as unstable?

Reply:
Yes, naming aside, that's basically the other model for this.

They should be public in the `provider` module with the yellow box warning people that the APIs may be unstable.

I would make the second one `pub struct UnvalidatedPluralCategory(pub u8)` that is a u8 for all purposes except that `impl Serialize` makes a nice string, returning an error if out of range, and `impl Deserialize` knows how to handle the string.

However I'd like to see what @robertbastian or @Manishearth think about these two approaches (re-use the existing enum and make it ZeroVec compatible, or introduce two new types which are data provider specific)

Reply:
> Yes, naming aside, that's basically the other model for this.
> 
> They should be public in the `provider` module with the yellow box warning people that the APIs may be unstable.
> 
> I would make the second one `pub struct UnvalidatedPluralCategory(pub u8)` that is a u8 for all purposes except that `is_human_readable impl Serialize` makes a nice string, returning an error if out of range, and `is_human_readable impl Deserialize` knows how to handle the string.
> 
> However I'd like to see what @robertbastian or @Manishearth think about these two approaches (re-use the existing enum and make it ZeroVec compatible, or introduce two new types which are data provider specific)

Sounds good! I've already finished implementing the API in another branch, so I'll wait for your consensus in the meantime.

Reply:
I agree that a new type is the best solution.

Reply:
Same